File: backend/libs/core/kactus-common/src/kactus_common/text.py
"""Text normalisation helpers (pure infrastructure — no domain knowledge)."""

from __future__ import annotations

import unicodedata

# ``đ``/``Đ`` is a distinct letter, not a base + combining mark, so NFD leaves it
# untouched and stripping marks alone would never make "duc" match "Đức".
_EXTRA_FOLDS = str.maketrans({"đ": "d", "Đ": "d", "Ð": "d"})


def normalize_search(value: str | None) -> str:
    """Fold a string to a diacritic-free, case-insensitive search key.

    Applied to **both** sides of a comparison so a Vietnamese name matches what a
    user types on an unaccented keyboard: ``normalize_search("Thành Đức")`` is
    ``"thanh duc"``, which ``"duc"`` is a substring of. Mirrors the frontend's
    ``normalizeText`` in ``src/lib/text.ts`` — keep the two in step.
    """
    if not value:
        return ""
    folded = value.translate(_EXTRA_FOLDS)
    decomposed = unicodedata.normalize("NFD", folded)
    without_marks = "".join(c for c in decomposed if not unicodedata.combining(c))
    return without_marks.casefold().strip()

File: frontend/packages/bloom-app/src/lib/text.ts
/** Text normalisation helpers for client-side search. */

// đ/Đ is a distinct letter, not a base + combining mark, so NFD leaves it alone
// and stripping marks would never make "duc" match "Đức".
const D_STROKE = /[đĐÐ]/g;
const DIACRITICS = /\p{Diacritic}/gu;

/**
 * Fold a string to a diacritic-free, case-insensitive search key.
 *
 * Apply it to **both** sides of a comparison so a Vietnamese name matches what
 * the user types on an unaccented keyboard: `normalizeText('Thành Đức')` is
 * `'thanh duc'`, which `'duc'` is a substring of. Mirrors the backend's
 * `normalize_search` in `kactus_common/text.py` — keep the two in step.
 */
export function normalizeText(value: string | null | undefined): string {
  if (!value) return '';
  return value.replace(D_STROKE, 'd').normalize('NFD').replace(DIACRITICS, '').toLowerCase().trim();
}

/** Whether `haystack` contains `needle`, ignoring case and diacritics. */
export function matchesSearch(haystack: string | null | undefined, needle: string): boolean {
  const q = normalizeText(needle);
  return q === '' || normalizeText(haystack).includes(q);
}
